File: app/src/main/java/com/example/recipeclipper/ui/mealtypes/MealTypesViewModel.kt
package com.example.recipeclipper.ui.mealtypes

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.recipeclipper.data.MealPlanRepository
import com.example.recipeclipper.data.model.MealType
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * [renaming] and [deleting] are the meal type whose rename field or delete confirmation is
 * open. [creating] is the "+ New meal type" field, expanded inline as on Lists.
 */
data class MealTypesUiState(
    val types: List<MealType> = emptyList(),
    val creating: Boolean = false,
    val newName: String = "",
    val renaming: MealType? = null,
    val renameText: String = "",
    val deleting: MealType? = null
)

/**
 * The Week menu's "Meal types" screen (#49): add, rename and reorder any, delete the user's
 * own. The types are collected here, not through `stateIn`, because moving one reads the
 * current order.
 */
@HiltViewModel
class MealTypesViewModel @Inject constructor(
    private val repository: MealPlanRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(MealTypesUiState())
    val uiState: StateFlow<MealTypesUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            repository.observeMealTypes().collect { types -> _uiState.update { it.copy(types = types) } }
        }
    }

    fun onStartCreating() = _uiState.update { it.copy(creating = true) }

    fun onNewNameChange(value: String) = _uiState.update { it.copy(newName = value) }

    fun onCancelCreating() = _uiState.update { it.copy(creating = false, newName = "") }

    fun onCreate() {
        val name = _uiState.value.newName.trim()
        if (name.isEmpty()) return
        _uiState.update { it.copy(creating = false, newName = "") }
        viewModelScope.launch { repository.addMealType(name) }
    }

    fun onRenameStart(type: MealType) = _uiState.update { it.copy(renaming = type, renameText = type.name) }

    fun onRenameTextChange(value: String) = _uiState.update { it.copy(renameText = value) }

    fun onRenameConfirm() {
        val type = _uiState.value.renaming ?: return
        val name = _uiState.value.renameText.trim()
        if (name.isEmpty()) return
        _uiState.update { it.copy(renaming = null, renameText = "") }
        viewModelScope.launch { repository.renameMealType(type.id, name) }
    }

    fun onRenameDismissed() = _uiState.update { it.copy(renaming = null, renameText = "") }

    fun onMoveUp(type: MealType) = move(type, -1)

    fun onMoveDown(type: MealType) = move(type, +1)

    private fun move(type: MealType, by: Int) {
        val order = _uiState.value.types.toMutableList()
        val from = order.indexOfFirst { it.id == type.id }
        val to = from + by
        if (from < 0 || to !in order.indices) return
        order.add(to, order.removeAt(from))
        // Shown at once; the database's echo agrees.
        _uiState.update { it.copy(types = order) }
        viewModelScope.launch { repository.reorderMealTypes(order.map { it.id }) }
    }

    /** Only the user's own types offer Delete; a seeded one is ignored here too. */
    fun onDeleteStart(type: MealType) {
        if (type.isBuiltIn) return
        _uiState.update { it.copy(deleting = type) }
    }

    fun onDeleteConfirm() {
        val type = _uiState.value.deleting ?: return
        _uiState.update { it.copy(deleting = null) }
        viewModelScope.launch { repository.deleteMealType(type.id) }
    }

    fun onDeleteDismissed() = _uiState.update { it.copy(deleting = null) }
}

File: ios/RecipeClipper/UI/MealTypes/MealTypesViewModel.swift
import Combine
import Foundation
import Observation

/// `renaming` and `deleting` are the meal type whose rename field or delete confirmation is
/// open. `creating` is the "+ New meal type" field, expanded inline as on Lists.
struct MealTypesUiState: Equatable {
    var types: [MealType] = []
    var creating = false
    var newName = ""
    var renaming: MealType?
    var renameText = ""
    var deleting: MealType?
}

/// The Week menu's "Meal types" screen (#49; Android's MealTypesViewModel): add, rename and
/// reorder any, delete the user's own (their meals move to Dinner).
@MainActor
@Observable
final class MealTypesViewModel {
    private(set) var uiState = MealTypesUiState()

    @ObservationIgnored private let repository: MealPlanRepository
    @ObservationIgnored private var typesSubscription: AnyCancellable?
    /// Writes run one after another, so two quick moves reach the database in the order made.
    @ObservationIgnored private var lastWrite: Task<Void, Never>?

    init(repository: MealPlanRepository) {
        self.repository = repository
        typesSubscription = repository.observeMealTypes()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] types in self?.uiState.types = types }
    }

    func onStartCreating() { uiState.creating = true }

    func onNewNameChange(_ value: String) { uiState.newName = value }

    func onCancelCreating() {
        uiState.creating = false
        uiState.newName = ""
    }

    func onCreate() {
        let name = uiState.newName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return }
        onCancelCreating()
        write { await $0.addMealType(name: name) }
    }

    func onRenameStart(_ type: MealType) {
        uiState.renaming = type
        uiState.renameText = type.name
    }

    func onRenameTextChange(_ value: String) { uiState.renameText = value }

    func onRenameConfirm() {
        guard let type = uiState.renaming else { return }
        let name = uiState.renameText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return }
        onRenameDismissed()
        write { await $0.renameMealType(id: type.id, name: name) }
    }

    func onRenameDismissed() {
        uiState.renaming = nil
        uiState.renameText = ""
    }

    func onMoveUp(_ type: MealType) { move(type, by: -1) }

    func onMoveDown(_ type: MealType) { move(type, by: 1) }

    private func move(_ type: MealType, by offset: Int) {
        var order = uiState.types
        guard let from = order.firstIndex(where: { $0.id == type.id }) else { return }
        let to = from + offset
        guard order.indices.contains(to) else { return }
        order.insert(order.remove(at: from), at: to)
        // Shown at once; the database's echo agrees.
        uiState.types = order
        let ids = order.map(\.id)
        write { await $0.reorderMealTypes(ids) }
    }

    /// Only the user's own types offer Delete; a seeded one is ignored here too.
    func onDeleteStart(_ type: MealType) {
        guard !type.isBuiltIn else { return }
        uiState.deleting = type
    }

    func onDeleteConfirm() {
        guard let type = uiState.deleting else { return }
        uiState.deleting = nil
        write { await $0.deleteMealType(id: type.id) }
    }

    func onDeleteDismissed() { uiState.deleting = nil }

    private func write(_ body: @escaping (MealPlanRepository) async -> Void) {
        let previous = lastWrite
        lastWrite = Task { [repository] in
            await previous?.value
            await body(repository)
        }
    }

    /// Waits for every queued write. For tests.
    func settleWrites() async { await lastWrite?.value }
}
